python/revng/daemon/event_manager.py

```diff
@@ -39,98 +26,27 @@ async def inner(*args, **kwargs):
-            for listener in self.handlers:
-                listener(self.manager, type_, self.get_token(headers))
-            for listener in self.cls_handlers:
-                listener(self.manager, type_, self.get_token(headers))
-
-        # Run the callbacks in an executor to avoid blocking the result
-        result = self.executor.submit(handler)
-        result.add_done_callback(self.handle_event_result)
-
-    @staticmethod
-    def get_token(headers: Headers) -> str:
-        if "authorization" not in headers or not headers["authorization"].startswith("Bearer "):
-            return ""
-        token = headers["authorization"][len("Bearer ") :]
-        if re.search(r"\s", token):
-            return ""
-        return token
-
-    @staticmethod
-    def handle_event_result(result_future: Future):
-        ex = result_future.exception()
-        if ex is not None:
-            logging.error("Exception raised while handling event callbacks", exc_info=ex)
-            signal.raise_signal(signal.SIGINT)
-
-
-class FIFOListener(EventHandler):
-    def __init__(self):
-        self.fifos: Dict[str, List[TextIOWrapper]] = {"begin": [], "context": []}
-        if "REVNG_NOTIFY_FIFOS" in os.environ:
-            fifo_string = os.environ["REVNG_NOTIFY_FIFOS"]
-            for fifo_definition in fifo_string.split(","):
-                fifo_path, fifo_type = fifo_definition.rsplit(":", 1)
-                if fifo_type in self.fifos:
-                    self.fifos[fifo_type].append(open(fifo_path, "w"))  # noqa: SIM115
-
-    def __call__(self, manager: Manager, type_: EventType, token: str):
-        if type_ == EventType.BEGIN:
-            begin_step = manager.get_step("begin")
-            assert begin_step is not None
-
-            tmpdir = mkdtemp()
-            begin_step.save(tmpdir)
-            self.write_fifos(type_, f"PUSH begin {tmpdir} {token}\n")
-
-        elif type_ == EventType.CONTEXT:
-            tmpdir = mkdtemp()
-            manager.save_context(tmpdir)
-            self.write_fifos(type_, f"PUSH context {tmpdir} {token}\n")
-
-    def write_fifos(self, type_: EventType, content: str):
-        if str(type_) not in self.fifos:
-            return
-
-        for fifo in self.fifos[str(type_)]:
-            try:
-                fifo.write(content)
-                fifo.flush()
-            except (BrokenPipeError, FileNotFoundError, PermissionError, TimeoutError) as e:
-                logging.warn(
-                    f"Encountered error when writing to notification pipe: {repr(e)}",
-                )
-
-    def __del__(self):
-        for fifos in self.fifos.values():
-            for fifo in fifos:
-                fifo.close()
-
-
-EventManager.add_class_handler(FIFOListener())
+        self.running = True
+        self.next_save: float | None = None
+
+    def run(self):
+        while self.running:
+            if self.next_save is not None and self.next_save < time.time():
+                self.manager.save()
```

Comment:
> Aren't we saving every time we change something?

That would be ultra-lagging: it takes a good 20s to `storeToDisk` (on a local setup, I'd say 2x conservatively when on the cloud). As [described in the pad](https://pad.rev.ng/99Bh2R02RNuL8uNugMv3-Q) I'd tweak a few things in pipeline before saving on every change.

> Why is this required?

So that saves are done only when the daemon is idling, so that the user does not "feel the lag"

Reply:
> What about after an analysis?

Both `runAnalysis` and `runAnalysesList` save everything (model and `Lift/module.ll`) if the diff map is not empty

Reply:
Added TODO